File: android/src/main/kotlin/dev/brewkits/native_workmanager/workers/FileDecompressionWorker.kt
package dev.brewkits.native_workmanager.workers

import android.util.Log
import dev.brewkits.kmpworkmanager.background.domain.AndroidWorker
import dev.brewkits.kmpworkmanager.background.domain.WorkerResult
import dev.brewkits.native_workmanager.workers.utils.SecurityValidator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.io.File
import java.io.FileOutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

/**
 * Native file decompression worker for Android.
 *
 * Extracts ZIP files using streaming to minimize memory usage.
 * Includes security protections against zip slip attacks.
 *
 * **Configuration JSON:**
 * ```json
 * {
 *   "zipPath": "/path/to/archive.zip",
 *   "targetDir": "/path/to/extract/",
 *   "overwrite": true,           // Optional: Overwrite existing files (default: true)
 *   "deleteAfterExtract": false  // Optional: Delete ZIP after successful extraction (default: false)
 * }
 * ```
 *
 * **Features:**
 * - Streaming extraction (low memory usage)
 * - Zip slip protection (prevents path traversal attacks)
 * - Auto-creates target directory
 * - Atomic operations (cleanup on error)
 * - Progress tracking support
 *
 * **Security:**
 * - Validates all extracted paths are within target directory
 * - Prevents path traversal via ".." in ZIP entries
 * - Validates file sizes during extraction
 *
 * **Performance:** ~5-10MB RAM regardless of ZIP size
 */
class FileDecompressionWorker : AndroidWorker {

    companion object {
        private const val TAG = "FileDecompressionWorker"
        private const val BUFFER_SIZE = 8192
        private const val MAX_ENTRY_COUNT = 10_000   // FS-M-002: zip with too many entries
    }

    data class Config(
        val zipPath: String,
        val targetDir: String,
        val overwrite: Boolean = true,
        val deleteAfterExtract: Boolean = false
    )

    override suspend fun doWork(input: String?): WorkerResult = withContext(Dispatchers.IO) {
        if (input.isNullOrEmpty()) {
            throw IllegalArgumentException("Input JSON is required")
        }

        // Parse configuration
        val config = try {
            val j = org.json.JSONObject(input)
            Config(
                zipPath = j.getString("zipPath"),
                targetDir = j.getString("targetDir"),
                overwrite = j.optBoolean("overwrite", true),
                deleteAfterExtract = j.optBoolean("deleteAfterExtract", false)
            )
        } catch (e: Exception) {
            throw IllegalArgumentException("Invalid config JSON: ${e.message}", e)
        }

        // Extract taskId for progress reporting
        val taskId = try {
            org.json.JSONObject(input).optString("__taskId", null)
        } catch (e: Exception) {
            null
        }

        // FIX H1: Use canonical-path validation (replaces bypassable contains("..") check)
        if (!SecurityValidator.validateFilePathSafe(config.zipPath)) {
            Log.e(TAG, "Error - Invalid or unsafe ZIP path")
            return@withContext WorkerResult.Failure("Invalid ZIP path")
        }

        if (!SecurityValidator.validateFilePathSafe(config.targetDir)) {
            Log.e(TAG, "Error - Invalid or unsafe target directory")
            return@withContext WorkerResult.Failure("Invalid target directory")
        }

        // Validate ZIP file exists
        val zipFile = File(config.zipPath)
        if (!zipFile.exists()) {
            Log.e(TAG, "Error - ZIP file not found: ${config.zipPath}")
            return@withContext WorkerResult.Failure("ZIP file not found: ${config.zipPath}")
        }

        if (!zipFile.isFile) {
            Log.e(TAG, "Error - Path is not a file: ${config.zipPath}")
            return@withContext WorkerResult.Failure("Path is not a file")
        }

        // ✅ SECURITY: Validate ZIP file size
        if (!SecurityValidator.validateFileSize(zipFile)) {
            Log.e(TAG, "Error - ZIP file too large")
            return@withContext WorkerResult.Failure("ZIP file exceeds size limit")
        }

        // Create target directory
        val targetDir = File(config.targetDir)
        if (!targetDir.exists()) {
            if (!targetDir.mkdirs()) {
                Log.e(TAG, "Error - Failed to create target directory")
                return@withContext WorkerResult.Failure("Failed to create target directory")
            }
            Log.d(TAG, "Created directory: ${targetDir.path}")
        }

        // Get canonical path for security checks
        val canonicalTargetPath = targetDir.canonicalPath

        Log.d(TAG, "Extracting ${zipFile.name} (${zipFile.length()} bytes)")
        Log.d(TAG, "Target: ${targetDir.path}")

        var extractedFiles = 0
        var extractedDirs = 0
        var totalBytes = 0L
        val extractedPaths = mutableListOf<String>()

        // Extract ZIP file
        return@withContext try {
            ZipInputStream(zipFile.inputStream()).use { zipInput ->
                var entry: ZipEntry? = zipInput.nextEntry
                var entryCount = 0   // FS-M-002: entry count guard

                while (entry != null) {
                    // FS-M-002: Prevent zip archives with excessive entry counts
                    entryCount++
                    if (entryCount > MAX_ENTRY_COUNT) {
                        Log.e(TAG, "Security - Entry count exceeded limit ($MAX_ENTRY_COUNT)")
                        cleanupExtractedFiles(extractedPaths)
                        return@withContext WorkerResult.Failure("Archive entry count exceeds limit ($MAX_ENTRY_COUNT)")
                    }

                    val entryName = entry.name

                    // ✅ SECURITY: Prevent zip slip attack
                    val destFile = File(targetDir, entryName)
                    val canonicalDestPath = destFile.canonicalPath

                    if (!canonicalDestPath.startsWith(canonicalTargetPath + File.separator)) {
                        Log.e(TAG, "Security - Zip slip attack detected: $entryName")
                        // Clean up already extracted files
                        cleanupExtractedFiles(extractedPaths)
                        return@withContext WorkerResult.Failure("Zip slip attack detected in entry: $entryName")
                    }

                    if (entry.isDirectory) {
                        // Create directory
                        if (!destFile.exists()) {
                            if (!destFile.mkdirs()) {
                                Log.e(TAG, "Error - Failed to create directory: $entryName")
                                cleanupExtractedFiles(extractedPaths)
                                return@withContext WorkerResult.Failure("Failed to create directory: $entryName")
                            }
                            extractedDirs++
                            extractedPaths.add(destFile.absolutePath)
                        }
                    } else {
                        // Extract file
                        // Check if file exists and overwrite setting
                        if (destFile.exists() && !config.overwrite) {
                            Log.d(TAG, "Skipping existing file: $entryName")
                            zipInput.closeEntry()
                            entry = zipInput.nextEntry
                            continue
                        }

                        // Create parent directories if needed
                        destFile.parentFile?.let { parent ->
                            if (!parent.exists() && !parent.mkdirs()) {
                                Log.e(TAG, "Error - Failed to create parent directory: ${parent.path}")
                                cleanupExtractedFiles(extractedPaths)
                                return@withContext WorkerResult.Failure("Failed to create parent directory")
                            }
                        }

                        // Extract file content
                        var bytesWritten = 0L
                        FileOutputStream(destFile).use { output ->
                            val buffer = ByteArray(BUFFER_SIZE)
                            var len: Int
                            while (zipInput.read(buffer).also { len = it } > 0) {
                                output.write(buffer, 0, len)
                                bytesWritten += len
                                totalBytes += len

                                // ✅ SECURITY: Robust Zip Bomb Protection
                                // Check 1: Absolute Total Size Limit (2GB)
                                if (totalBytes > 2L * 1024 * 1024 * 1024) {
                                    Log.e(TAG, "Security - Total extracted size exceeded limit (2GB)")
                                    cleanupExtractedFiles(extractedPaths)
                                    destFile.delete()  // FS-H-003: delete partial current file
                                    return@withContext WorkerResult.Failure("Total extracted size too large (Zip Bomb protection)")
                                }

                                // Check 2: Dynamic Compression Ratio (Max 100:1)
                                if (entry.size > 0 && bytesWritten > entry.size * 100) {
                                    Log.e(TAG, "Security - Suspicious compression ratio detected (>100:1) for: $entryName")
                                    cleanupExtractedFiles(extractedPaths)
                                    destFile.delete()  // FS-H-003: delete partial current file
                                    return@withContext WorkerResult.Failure("Suspicious compression ratio (Zip Bomb protection)")
                                }
                            }
                        }

                        extractedFiles++
                        extractedPaths.add(destFile.absolutePath)

                        // Set file modification time to match ZIP entry
                        if (entry.time > 0) {
                            destFile.setLastModified(entry.time)
                        }

                        Log.d(TAG, "Extracted: $entryName (${bytesWritten} bytes)")
                    }

                    zipInput.closeEntry()
                    entry = zipInput.nextEntry
                }
            }

            // ✅ Optional: Delete ZIP file after successful extraction
            if (config.deleteAfterExtract) {
                if (zipFile.delete()) {
                    Log.d(TAG, "Deleted ZIP file: ${zipFile.name}")
                } else {
                    Log.w(TAG, "Failed to delete ZIP file: ${zipFile.name}")
                }
            }

            Log.d(TAG, "Extraction complete")
            Log.d(TAG, "  Files: $extractedFiles, Directories: $extractedDirs")
            Log.d(TAG, "  Total size: $totalBytes bytes")

            // ✅ Return success with extraction data
            WorkerResult.Success(
                message = "Extracted $extractedFiles files, $extractedDirs directories",
                data = buildJsonObject {
                    put("extractedFiles", extractedFiles)
                    put("extractedDirs", extractedDirs)
                    put("totalBytes", totalBytes)
                    put("targetDir", targetDir.absolutePath)
                    put("zipDeleted", config.deleteAfterExtract)
                }
            )

        } catch (e: Exception) {
            Log.e(TAG, "Error during extraction: ${e.message}", e)
            // Clean up partially extracted files on error
            cleanupExtractedFiles(extractedPaths)
            WorkerResult.Failure(
                message = e.message ?: "Unknown error during extraction",
                shouldRetry = false
            )
        }
    }

    /**
     * Clean up extracted files in case of error.
     */
    private fun cleanupExtractedFiles(paths: List<String>) {
        Log.d(TAG, "Cleaning up ${paths.size} extracted files/directories...")
        paths.reversed().forEach { path ->
            try {
                val file = File(path)
                if (file.exists()) {
                    if (file.isDirectory) {
                        file.deleteRecursively()
                    } else {
                        file.delete()
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "Failed to cleanup: $path - ${e.message}")
            }
        }
    }
}

File: ios/Classes/workers/FileDecompressionWorker.swift
import Foundation
import Compression
import ZIPFoundation

/// Native file decompression worker for iOS.
///
/// Extracts ZIP files using streaming to minimize memory usage.
/// Includes security protections against zip slip attacks.
///
/// **Configuration JSON:**
/// ```json
/// {
///   "zipPath": "/path/to/archive.zip",
///   "targetDir": "/path/to/extract/",
///   "overwrite": true,           // Optional: Overwrite existing files (default: true)
///   "deleteAfterExtract": false  // Optional: Delete ZIP after successful extraction (default: false)
/// }
/// ```
///
/// **Features:**
/// - Streaming extraction (low memory usage)
/// - Zip slip protection (prevents path traversal attacks)
/// - Auto-creates target directory
/// - Atomic operations (cleanup on error)
///
/// **Security:**
/// - Validates all extracted paths are within target directory
/// - Prevents path traversal via ".." in ZIP entries
/// - Validates file sizes during extraction
///
/// **Performance:** ~5-10MB RAM regardless of ZIP size
class FileDecompressionWorker: IosWorker {

    struct Config: Codable {
        let zipPath: String
        let targetDir: String
        let overwrite: Bool?
        let deleteAfterExtract: Bool?

        var shouldOverwrite: Bool {
            overwrite ?? true
        }

        var shouldDeleteZip: Bool {
            deleteAfterExtract ?? false
        }
    }

    func doWork(input: String?) async throws -> WorkerResult {
        guard let input = input, !input.isEmpty else {
            print("FileDecompressionWorker: Error - Empty or null input")
            return .failure(message: "Empty or null input")
        }

        // Parse configuration
        let config: Config
        do {
            let data = input.data(using: .utf8)!
            config = try JSONDecoder().decode(Config.self, from: data)
        } catch {
            print("FileDecompressionWorker: Error parsing JSON config: \(error)")
            return .failure(message: "Invalid JSON config: \(error.localizedDescription)")
        }

        // ✅ SECURITY: Validate paths
        guard SecurityValidator.validateFilePath(config.zipPath) else {
            print("FileDecompressionWorker: Error - Invalid ZIP path")
            return .failure(message: "Invalid ZIP path")
        }

        guard SecurityValidator.validateFilePath(config.targetDir) else {
            print("FileDecompressionWorker: Error - Invalid target directory")
            return .failure(message: "Invalid target directory")
        }

        let zipURL = URL(fileURLWithPath: config.zipPath)
        let targetDirURL = URL(fileURLWithPath: config.targetDir)

        // Validate ZIP file exists
        guard FileManager.default.fileExists(atPath: zipURL.path) else {
            print("FileDecompressionWorker: Error - ZIP file not found: \(config.zipPath)")
            return .failure(message: "ZIP file not found")
        }

        // ✅ SECURITY: Validate archive size
        guard SecurityValidator.validateArchiveSize(zipURL) else {
            print("FileDecompressionWorker: Error - Archive size exceeds limit")
            return .failure(message: "Archive size exceeds limit")
        }

        // Create target directory if needed
        if !FileManager.default.fileExists(atPath: targetDirURL.path) {
            do {
                try FileManager.default.createDirectory(at: targetDirURL,
                                                       withIntermediateDirectories: true)
                print("FileDecompressionWorker: Created directory: \(targetDirURL.path)")
            } catch {
                print("FileDecompressionWorker: Error creating directory: \(error)")
                return .failure(message: "Failed to create target directory: \(error.localizedDescription)")
            }
        }

        // Get canonical path for security checks
        let canonicalTargetPath = targetDirURL.standardizedFileURL.path

        print("FileDecompressionWorker: Extracting \(zipURL.lastPathComponent)")
        print("  Target: \(targetDirURL.path)")

        var extractedFiles = 0
        var extractedDirs = 0
        var totalBytes: Int64 = 0
        var extractedPaths: [String] = []

        do {
            // Use Foundation's built-in unzipping
            let fileManager = FileManager.default

            // For iOS, we'll use a simpler approach with NSFileCoordinator and unzipping
            // Note: iOS doesn't have built-in ZIP support in Foundation, so we use a workaround

            // Create a temporary directory for extraction
            let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
            try fileManager.createDirectory(at: tempDir, withIntermediateDirectories: true)

            defer {
                // Cleanup temp dir
                try? fileManager.removeItem(at: tempDir)
            }

            // iOS Workaround: Use unzip command via Process
            // This is a simplified implementation. In production, consider using:
            // - ZIPFoundation library (https://github.com/weichsel/ZIPFoundation)
            // - Compression framework for more control

            // For now, we'll manually iterate through ZIP entries using basic file operations
            // This is a placeholder that should be replaced with proper ZIP library

            let unzipResult = try await unzipFile(
                zipURL: zipURL,
                destinationURL: targetDirURL,
                canonicalPath: canonicalTargetPath,
                overwrite: config.shouldOverwrite
            )

            extractedFiles = unzipResult.files
            extractedDirs = unzipResult.dirs
            totalBytes = unzipResult.bytes
            extractedPaths = unzipResult.paths

            // ✅ Optional: Delete ZIP file after successful extraction
            if config.shouldDeleteZip {
                try fileManager.removeItem(at: zipURL)
                print("FileDecompressionWorker: Deleted ZIP file")
            }

            print("FileDecompressionWorker: Extraction complete")
            print("  Files: \(extractedFiles), Directories: \(extractedDirs)")
            print("  Total size: \(totalBytes) bytes")

            // ✅ Return success with extraction data
            return .success(
                message: "Extracted \(extractedFiles) files, \(extractedDirs) directories",
                data: [
                    "extractedFiles": extractedFiles,
                    "extractedDirs": extractedDirs,
                    "totalBytes": totalBytes,
                    "targetDir": targetDirURL.path,
                    "zipDeleted": config.shouldDeleteZip
                ]
            )

        } catch {
            print("FileDecompressionWorker: Error during extraction: \(error)")
            // Clean up partially extracted files
            cleanupExtractedFiles(paths: extractedPaths)
            return .failure(message: error.localizedDescription)
        }
    }

    /// Unzip file using NSTask/Process (iOS workaround)
    /// Unzip file using ZIPFoundation library.
    private func unzipFile(
        zipURL: URL,
        destinationURL: URL,
        canonicalPath: String,
        overwrite: Bool
    ) async throws -> (files: Int, dirs: Int, bytes: Int64, paths: [String]) {
        let fileManager = FileManager.default
        var extractedFiles = 0
        var extractedDirs = 0
        var totalBytes: Int64 = 0
        var paths: [String] = []

        guard let archive = Archive(url: zipURL, accessMode: .read) else {
            throw NSError(domain: "FileDecompressionWorker", code: -1,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to open ZIP archive"])
        }

        for entry in archive {
            let destinationEntryURL = destinationURL.appendingPathComponent(entry.path)

            // ✅ SECURITY: Zip Slip Protection - ensure extracted path is within target directory
            let canonicalDestinationEntryPath = destinationEntryURL.standardizedFileURL.path
            guard canonicalDestinationEntryPath.hasPrefix(canonicalPath) else {
                throw NSError(
                    domain: "FileDecompressionWorker",
                    code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "Zip slip attack detected for entry: \(entry.path)"]
                )
            }

            // Handle overwrite logic
            if fileManager.fileExists(atPath: destinationEntryURL.path) {
                if !overwrite {
                    print("FileDecompressionWorker: Skipping existing file (overwrite disabled): \(entry.path)")
                    continue
                } else {
                    // Attempt to remove existing item for overwrite
                    try? fileManager.removeItem(at: destinationEntryURL)
                }
            }

            // Create parent directories if needed
            let parentDirectory = destinationEntryURL.deletingLastPathComponent()
            if !fileManager.fileExists(atPath: parentDirectory.path) {
                try fileManager.createDirectory(at: parentDirectory, withIntermediateDirectories: true)
            }

            // Extract entry
            print("FileDecompressionWorker: Extracting \(entry.path)")
            try archive.extract(entry, to: destinationEntryURL, skipCRC32: false)

            paths.append(destinationEntryURL.path)

            // Update stats
            if entry.type == .directory {
                extractedDirs += 1
            } else {
                extractedFiles += 1
                totalBytes += Int64(entry.uncompressedSize)
            }
        }
        return (extractedFiles, extractedDirs, totalBytes, paths)
    }

    /// Clean up extracted files in case of error
    private func cleanupExtractedFiles(paths: [String]) {
        print("FileDecompressionWorker: Cleaning up \(paths.count) extracted files/directories...")

        // Reverse order to delete files before their parent directories
        for path in paths.reversed() {
            do {
                try FileManager.default.removeItem(atPath: path)
            } catch {
                print("FileDecompressionWorker: Failed to cleanup \(path): \(error)")
            }
        }
    }
}
